core/src/scene/asset.cpp

```diff
@@ -0,0 +1,135 @@
+    std::unique_ptr<mz_zip_archive> archiveHandle;
+    std::map<std::string, unsigned int> fileIndices = {};
+
+    std::vector<char> data;
+};
+
+Asset::Asset(std::string name, std::string path, std::shared_ptr<ZipHandle> zipHandle,
+        std::vector<char> zippedData) :
+        m_name(name),
+        m_path(path),
+        m_zipHandle(zipHandle) {
+
+    buildZipHandle(zippedData);
+}
+
+Asset::~Asset() {
+    if (m_zipHandle && m_zipHandle->archiveHandle) {
+        mz_zip_reader_end(m_zipHandle->archiveHandle.get());
+    }
+}
+
+bool Asset::isBaseYaml(const std::string& filename) {
+    auto extLoc = filename.find(".yaml");
+    if (extLoc == std::string::npos) { return false; }
+
+    auto slashLoc = filename.find("/");
+    if (slashLoc != std::string::npos) { return false; }
+    return true;
+}
+
+void Asset::buildZipHandle(const std::vector<char>& zipData) {
+
+    if (zipData.empty()) { return; }
+
+    m_zipHandle = std::make_shared<ZipHandle>();
+    m_zipHandle->archiveHandle.reset(new mz_zip_archive());
+    m_zipHandle->data = zipData;
+
+    mz_zip_archive* zip = m_zipHandle->archiveHandle.get();
+    memset(zip, 0, sizeof(mz_zip_archive));
+    if (!mz_zip_reader_init_mem(zip, m_zipHandle->data.data(), m_zipHandle->data.size(), 0)) {
+        LOGE("ZippedAssetPackage: Could not open archive: %s", m_name.c_str());
+        m_zipHandle.reset();
+        return;
+    }
+
+    /* Instead of using mz_zip_reader_locate_file, maintaining a map of file name to index,
+     * for performance reasons.
+     * https://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/util/compress/api/miniz/miniz.c
+     */
+    for (unsigned int i = 0; i < mz_zip_reader_get_num_files(zip); i++) {
+        mz_zip_archive_file_stat st;
+        if (!mz_zip_reader_file_stat(zip, i, &st)) {
+            LOGE("ZippedAssetPackage: Could not read file stats: %s", st.m_filename);
+            continue;
+        }
+        if (isBaseYaml(st.m_filename)) {
+            m_path = st.m_filename;
+        }
+        m_zipHandle->fileIndices[st.m_filename] = i;
+    }
+}
+
+
+std::vector<char> Asset::readBytesFromAsset(const std::shared_ptr<Platform>& platform, const std::string& filename) {
+
+    std::vector<char> fileData;
+
+    if (m_zipHandle) {
+        if (m_zipHandle->archiveHandle) {
+            mz_zip_archive* zip = m_zipHandle->archiveHandle.get();
+            auto it = m_zipHandle->fileIndices.find(filename);
+
+            if (it != m_zipHandle->fileIndices.end()) {
+                std::size_t elementSize = 0;
+                char* elementData = static_cast<char*>(mz_zip_reader_extract_to_heap(zip, it->second, &elementSize, 0));
+                if (!elementData) {
+                    LOGE("ZippedAssetPackage::loadAsset: Could not load archive asset: %s", filename.c_str());
+                } else {
+                    fileData.resize(elementSize);
+                    fileData.assign(elementData, elementData + elementSize);
```

Comment:
Addressed: a81f3d8a79132ad00c645bdbf1edcfb4f91f14fc

Now instead of using `mz_zip_reader_extract_to_heap`, I switched to use `mz_zip_reader_extract_to_mem` which allows to pass in a buffer to be filled with data.